File: call.py
import sys
import pyml

sys.path = ['', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-linux2', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages']


def pass_integer(n):
    return n


def test_raise(n):
    raise Exception("pyml_test_exception")


def test_custom_exception():
    try:
        pyml.call("test_raise", [0])
    except:
        return 1


def test_exception():
    try:
        pyml.call("divide_by", [0])
    except:
        return 1


def divide_by(n):
    return 5 / n


def divide(x, y):
    return x / y


def test_n_args():
    return pyml.call("divide", [10, 2])


def test_call():
    return pyml.call("pass_integer", [10])


def test_call_with_double():
    return int(pyml.call("fdivide", [5.5, 1.2]))


def test_call_with_string():
    return pyml.call("strcmp", ["UROP", "UROP"])


def test_call_with_bool():
    return pyml.call("boolcmp", [True, False])


def strcmp(x, y):
    return int(x == y)


def boolcmp(x, y):
    return x == y

File: call.ml
external call_python: string -> 'a list -> 'b = "call_python"

exception Pyml_exception of string;;

let pass_integer n = n;;

let test_raise n = raise(Pyml_exception "pyml_test_exception");;

let test_custom_exception () =
    try call_python "test_raise" [0] with
        Pyml_exception msg -> 1
;;

let divide_by n = 5 / n;;

let test_exception () =
    try call_python "divide_by" [0] with
        Pyml_exception msg -> 1
;;

let test_call () = call_python "pass_integer" [10];;

let divide x y = x / y;;

let fdivide x y = x /. y;;

let test_n_args () = call_python "divide" [10;2];;

let strcmp x y = String.compare x y;;

let boolcmp x y = x == y;;

let test_call_with_double () = int_of_float(call_python "divide" [5.5;1.2]);;

let test_call_with_string () = call_python "strcmp" ["UROP";"UROP"];;

let test_call_with_bool () = call_python "boolcmp" [true;false];;

let () =
    Callback.register "pass_integer" pass_integer;
    Callback.register "test_call" test_call;
    Callback.register "test_raise" test_raise;
    Callback.register "test_exception" test_exception;
    Callback.register "test_custom_exception" test_custom_exception;
    Callback.register "divide_by" divide_by;
    Callback.register "Printexc.to_string" Printexc.to_string;
    Callback.register "divide" divide;
    Callback.register "fdivide" fdivide;
    Callback.register "strcmp" strcmp;
    Callback.register "boolcmp" boolcmp;
    Callback.register "test_n_args" test_n_args;
    Callback.register "test_call_with_double" test_call_with_double;
    Callback.register "test_call_with_string" test_call_with_string;
    Callback.register "test_call_with_bool" test_call_with_bool;
    Callback.register_exception "pyml_exception" (Pyml_exception "pyml_test_exception");
;;
